toolchain/check/convert.h

```diff
@@ -0,0 +1,93 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_CHECK_CONVERT_H_
+#define CARBON_TOOLCHAIN_CHECK_CONVERT_H_
+
+#include "toolchain/check/context.h"
+#include "toolchain/check/pending_block.h"
+#include "toolchain/parse/tree.h"
+#include "toolchain/sem_ir/node.h"
+
+namespace Carbon::Check {
+
+// Description of the target of a conversion.
+struct ConversionTarget {
+  enum Kind {
+    // Convert to a value of type `type`.
+    Value,
+    // Convert to either a value or a reference of type `type`.
+    ValueOrReference,
+    // The result of the conversion is discarded. It can't be an initializing
+    // expression, but can be anything else.
+    Discarded,
+    // Convert to an initializer for the object denoted by `init_id`.
+    Initializer,
+    // Convert to an initializer for the object denoted by `init_id`,
+    // including a final destination store if needed.
+    FullInitializer,
+    Last = FullInitializer
+  };
+  // The kind of the target for this conversion.
+  Kind kind;
+  // The target type for the conversion.
+  SemIR::TypeId type_id;
+  // For an initializer, the object being initialized.
+  SemIR::NodeId init_id = SemIR::NodeId::Invalid;
+  // For an initializer, a block of pending instructions that are needed to
+  // form the value of `target_id`, and that can be discarded if no
+  // initialization is needed.
+  PendingBlock* init_block = nullptr;
+
+  // Are we converting this value into an initializer for an object?
+  bool is_initializer() const {
+    return kind == Initializer || kind == FullInitializer;
+  }
+};
+
+// Convert a value to another type and expression category.
+auto Convert(Context& context, Parse::Node parse_node, SemIR::NodeId value_id,
+             ConversionTarget target) -> SemIR::NodeId;
+
+// Performs initialization of `target_id` from `value_id`. Returns the
+// possibly-converted initialization expression, which should be assigned to
+// the target using a suitable node for the kind of initialization.
+auto Initialize(Context& context, Parse::Node parse_node,
+                SemIR::NodeId target_id, SemIR::NodeId value_id)
+    -> SemIR::NodeId;
```

Comment:
WDYT about this living in `initializing_expression.h`? (Potentially renaming it to `initialize.h` if needed, but given that this returns the initializing expression, the current name seems fine.

Or if that would create cycles, maybe there should just be one header/source for all of this called `initialize.h` or the current `convert.h`?

Also, minor tweak I think:
```suggestion
// Performs initialization of `target_id` from `value_id`. Returns the
// possibly-converted initializing expression, which should be assigned to
// the target using a suitable node for the kind of initialization.
auto Initialize(Context& context, Parse::Node parse_node,
                SemIR::NodeId target_id, SemIR::NodeId value_id)
    -> SemIR::NodeId;
```

Reply:
Applied the comment fix and inlined `initializing_expression.cpp` into `convert.cpp`.